File: RollnWrite/Games/QwixxMixx/MixxGame.swift
//
//  MixxGame.swift
//  RollnWrite – Qwixx Mixx
//
//  The Qwixx "gemixxt" (Mixx) engine: holds the state of BOTH boards (Variant A
//  and Variant B), enforces the per-row rules, and computes the score through an
//  injected `ScoringStrategy`.
//
//  SOLID notes:
//  - SRP: owns rules + state transitions only; scoring math is delegated
//         (`ScoringStrategy`), presentation lives in the view; the printed
//         layout lives in `MixxLayout`.
//  - DIP: the scoring strategy is injected (classic Qwixx cap 12).
//  - LSP: conforms to the generic `Scoreboard` protocol used by host UI.
//
//  Both boards share the classic Qwixx rule set (rows crossed strictly
//  left-to-right; the right-most cell locks the row but only after ≥5 earlier
//  crosses; four −5 penalties; game ends at two locks or the 4th penalty). They
//  differ only in their printed cell layout, so one engine serves both. The two
//  boards keep independent state and persist under separate keys, and `undo`,
//  `reset` and the score apply to the *currently selected* board.
//

import SwiftUI

@MainActor
public final class MixxGame: ObservableObject, Scoreboard {

    /// The board currently shown / acted upon.
    @Published public var board: MixxBoard {
        didSet { saveBoardSelection() }
    }

    @Published private var stateA = MixxState()
    @Published private var stateB = MixxState()

    /// Actions undone via `undo()`, most-recently-undone last, so `redo()` can
    /// re-apply them in LIFO order — one stack per board, mirroring `stateA`/
    /// `stateB`, since undo/redo/reset apply to the *currently selected* board.
    /// Deliberately NOT persisted (`MixxState` / `Codable` is untouched) and
    /// NOT part of either state — redo is an in-memory, per-session
    /// convenience, like most editors. Any new forward move (via
    /// `recordAction`) clears the current board's stack.
    private var redoStackA: [MixxAction] = []
    private var redoStackB: [MixxAction] = []

    /// Redo stack of the board currently in play — mirrors `state`.
    private var redoStack: [MixxAction] {
        get { board == .variantA ? redoStackA : redoStackB }
        set {
            if board == .variantA { redoStackA = newValue } else { redoStackB = newValue }
        }
    }

    private let scoring: ScoringStrategy
    private let persistencePrefix: String

    /// Classic Qwixx scoring: up to 12 valued crosses per row (78 points).
    public init(
        scoring: ScoringStrategy = TriangularScoring(cap: 12),
        persistencePrefix: String = "rollnwrite.qwixx.mixx"
    ) {
        self.scoring = scoring
        self.persistencePrefix = persistencePrefix
        self.board = .variantA
        load()
    }

    // MARK: - Current board state

    /// State of the board currently in play.
    private var state: MixxState {
        get { board == .variantA ? stateA : stateB }
        set {
            if board == .variantA { stateA = newValue } else { stateB = newValue }
        }
    }

    /// The printed layout of the current board.
    public var layout: [MixxRowLayout] { MixxLayout.rows(for: board) }

    // MARK: - Accessors

    public func rowState(_ rowIndex: Int) -> MixxRow {
        state.rows[rowIndex]
    }

    public func rowLayout(_ rowIndex: Int) -> MixxRowLayout {
        layout[rowIndex]
    }

    public var penalties: Int { state.penalties }

    public var lockedRowCount: Int {
        state.rows.filter { $0.locked }.count
    }

    // MARK: - Rule enforcement

    /// Whether crossing `index` in row `rowIndex` is a legal move right now.
    ///
    /// Enforces: game not over · row not locked · cell not already marked ·
    /// left-to-right · the right-most cell needs ≥5 earlier crosses to lock.
    public func canMark(_ rowIndex: Int, _ index: Int) -> Bool {
        guard !isGameOver else { return false }
        let r = state.rows[rowIndex]
        guard !r.locked, !r.marks.contains(index), index > r.maxMarkedIndex else { return false }
        if index == MixxRow.lockIndex {
            return r.marks.count >= 5
        }
        return true
    }

    public func mark(_ rowIndex: Int, _ index: Int) {
        guard canMark(rowIndex, index) else { return }
        var r = state.rows[rowIndex]
        r.marks.insert(index)
        var didLock = false
        if index == MixxRow.lockIndex {
            r.locked = true
            didLock = true
        }
        var s = state
        s.rows[rowIndex] = r
        recordAction(&s, .mark(row: rowIndex, index: index, didLock: didLock))
        state = s
        save()
    }

    // MARK: - Penalties

    public func canAddPenalty() -> Bool {
        !isGameOver && state.penalties < MixxState.maxPenalties
    }

    public func addPenalty() {
        guard canAddPenalty() else { return }
        var s = state
        s.penalties += 1
        recordAction(&s, .penalty)
        state = s
        save()
    }

    // MARK: - Concede a colour / finish manually

    /// You may close (concede) a row whose colour another player locked: the row
    /// closes for you, but you score no lock bonus — you never crossed its final
    /// cell. Allowed on any still-open row while the game is live.
    public func canConcedeRow(_ rowIndex: Int) -> Bool {
        !isGameOver && !state.rows[rowIndex].locked
    }

    public func concedeRow(_ rowIndex: Int) {
        guard canConcedeRow(rowIndex) else { return }
        var s = state
        s.rows[rowIndex].locked = true
        recordAction(&s, .concede(row: rowIndex))
        state = s
        save()
    }

    public func isLastConcede(_ rowIndex: Int) -> Bool {
        if case let .concede(r) = state.history.last { return r == rowIndex }
        return false
    }

    /// End the game by hand — e.g. another player crossed the final lock.
    public var canFinishManually: Bool { !isGameOver }

    public func finishGame() {
        guard canFinishManually else { return }
        var s = state
        s.manuallyFinished = true
        recordAction(&s, .finish)
        state = s
        save()
    }

    // MARK: - Scoreboard

    /// Crosses counted toward a row's score: its own marks plus the lock.
    public func crosses(_ rowIndex: Int) -> Int {
        state.rows[rowIndex].scoringCrosses
    }

    public func points(_ rowIndex: Int) -> Int {
        scoring.points(forCrosses: crosses(rowIndex))
    }

    public var penaltyPoints: Int { state.penalties * 5 }

    public var totalScore: Int {
        (0..<state.rows.count).reduce(0) { $0 + points($1) } - penaltyPoints
    }

    /// Ends when two rows are locked, the 4th penalty is taken, or the player
    /// ends it by hand.
    public var isGameOver: Bool {
        state.manuallyFinished || lockedRowCount >= 2 || state.penalties >= MixxState.maxPenalties
    }

    public var canUndo: Bool { !state.history.isEmpty }

    // MARK: - Tap-to-undo (strictly LIFO: only the most-recent action qualifies)

    /// `true` if the most-recent action is crossing `index` in row `rowIndex`,
    /// so tapping that cell un-checks it (a second way to undo).
    public func isLastMark(_ rowIndex: Int, _ index: Int) -> Bool {
        if case let .mark(r, i, _) = state.history.last { return r == rowIndex && i == index }
        return false
    }

    /// `true` if the most-recent action is the last penalty, so tapping it undoes it.
    public func isLastPenalty() -> Bool {
        if case .penalty = state.history.last { return true }
        return false
    }

    /// Reverse the most recent action on the current board. Strictly LIFO.
    public func undo() {
        var s = state
        guard let last = s.history.popLast() else { return }
        switch last {
        case let .mark(rowIndex, index, didLock):
            var r = s.rows[rowIndex]
            r.marks.remove(index)
            if didLock { r.locked = false }
            s.rows[rowIndex] = r
        case .penalty:
            s.penalties = max(0, s.penalties - 1)
        case let .concede(rowIndex):
            s.rows[rowIndex].locked = false
        case .finish:
            s.manuallyFinished = false
        }
        state = s
        redoStack.append(last)
        save()
    }

    public var canRedo: Bool { !redoStack.isEmpty }

    /// `true` while `redo()` is re-applying an action through its original
    /// mutator, so `recordAction` (called by that mutator) knows NOT to treat
    /// it as a fresh move and clear the rest of the current board's redo stack.
    private var isRedoing = false

    /// Re-apply the most recently undone action (on the current board) through
    /// the SAME mutator a fresh move takes, so scores/locks/derived state stay
    /// exact — never re-implement the effect here.
    public func redo() {
        guard let next = redoStack.popLast() else { return }
        isRedoing = true
        switch next {
        case let .mark(rowIndex, index, _):
            mark(rowIndex, index)
        case .penalty:
            addPenalty()
        case let .concede(rowIndex):
            concedeRow(rowIndex)
        case .finish:
            finishGame()
        }
        isRedoing = false
    }

    /// Clears only the currently selected board.
    public func reset() {
        state = MixxState()
        redoStack = []
        save()
    }

    // MARK: - Mutation helpers

    /// Appends a new action to `s`'s history. Any FORWARD move — i.e. every
    /// call site except `redo()` re-applying an undone one — invalidates the
    /// current board's redo stack (standard editor semantics: making a new
    /// move after undoing forecloses the redone future).
    private func recordAction(_ s: inout MixxState, _ action: MixxAction) {
        s.history.append(action)
        if !isRedoing { redoStack = [] }
    }

    // MARK: - Persistence

    private func stateKey(_ b: MixxBoard) -> String {
        "\(persistencePrefix).\(b.rawValue).state"
    }

    private var boardSelectionKey: String { "\(persistencePrefix).board" }

    private func save() {
        let key = stateKey(board)
        if let data = try? JSONEncoder().encode(state) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }

    private func saveBoardSelection() {
        UserDefaults.standard.set(board.rawValue, forKey: boardSelectionKey)
    }

    private func load() {
        if let data = UserDefaults.standard.data(forKey: stateKey(.variantA)),
           let restored = try? JSONDecoder().decode(MixxState.self, from: data) {
            stateA = restored
        }
        if let data = UserDefaults.standard.data(forKey: stateKey(.variantB)),
           let restored = try? JSONDecoder().decode(MixxState.self, from: data) {
            stateB = restored
        }
        if let raw = UserDefaults.standard.string(forKey: boardSelectionKey),
           let restored = MixxBoard(rawValue: raw) {
            board = restored
        }
    }
}

File: android/engine/src/main/kotlin/dev/bo3/rollnwrite/engine/mixx/MixxGame.kt
package dev.bo3.rollnwrite.engine.mixx

import dev.bo3.rollnwrite.engine.ScoringStrategy
import dev.bo3.rollnwrite.engine.Scoreboard
import dev.bo3.rollnwrite.engine.TriangularScoring

/**
 * The Qwixx "gemixxt" (Mixx) engine: holds the state of ONE board (the
 * caller picks Variant A or Variant B at construction time; switching boards
 * means constructing/holding a second instance - see [MixxViewModel] on the
 * Android app layer, mirroring iOS's `stateA`/`stateB` pair), enforces the
 * per-row rules, and computes the score through an injected [ScoringStrategy].
 *
 * Mirrors `RollnWrite/Games/QwixxMixx/MixxGame.swift` name-for-name (Kotlin
 * conventions), adapted to Kotlin's per-instance-board split: the Swift
 * engine owns BOTH boards' state in one object and switches via `board`:
 * `didSet`; this port keeps ONE board per engine instance (matching
 * `QwixxGame`'s pure-JVM, no-Android, no-persistence-in-the-engine shape) so
 * app-layer persistence (two independent `SharedPreferences` keys, one per
 * board) stays a thin, explicit wrapper rather than living in the engine.
 *
 * Both boards share the classic Qwixx rule set (rows crossed strictly
 * left-to-right; the right-most cell locks the row but only after >=5 earlier
 * crosses; four -5 penalties; game ends at two locks or the 4th penalty).
 * They differ only in their printed cell layout ([MixxLayout]), so one engine
 * class serves both - `board` just selects which layout `rowLayout` reports.
 *
 * Persistence lives OUTSIDE the engine (an Android app-layer concern): this
 * class exposes [state] as a get-only snapshot and [restore] to load one back
 * in, but never touches `SharedPreferences`/files itself.
 */
class MixxGame(
    val board: MixxBoard = MixxBoard.VARIANT_A,
    /** Classic Qwixx scoring: up to 12 valued crosses per row (78 points). */
    private val scoring: ScoringStrategy = TriangularScoring(cap = 12),
) : Scoreboard {

    var state: MixxState = MixxState()
        private set

    /**
     * Actions undone via [undo], most-recently-undone last, so [redo] can
     * re-apply them in LIFO order. Deliberately NOT part of [state] (kept
     * outside so it is never serialised) - redo is an in-memory,
     * per-session convenience. Any new forward move (via `recordAction`)
     * clears it.
     */
    private var redoStack: List<MixxAction> = emptyList()

    /**
     * `true` while [redo] is re-applying an action through its original
     * mutator, so `recordAction` (called by that mutator) knows NOT to
     * treat it as a fresh move and clear the rest of the redo stack.
     */
    private var isRedoing = false

    // --- Layout ---

    /** The printed layout of this engine's board. */
    val layout: List<MixxRowLayout> get() = MixxLayout.rows(board)

    // --- Accessors ---

    fun rowState(rowIndex: Int): MixxRow = state.rows[rowIndex]

    fun rowLayout(rowIndex: Int): MixxRowLayout = layout[rowIndex]

    val penalties: Int get() = state.penalties

    val lockedRowCount: Int get() = state.rows.count { it.locked }

    // --- Rule enforcement ---

    /**
     * Whether crossing [index] in row [rowIndex] is a legal move right now.
     *
     * Enforces: game not over, row not locked, cell not already marked,
     * left-to-right, and the right-most cell needs >=5 earlier crosses to
     * lock.
     */
    fun canMark(rowIndex: Int, index: Int): Boolean {
        if (isGameOver) return false
        val r = state.rows[rowIndex]
        if (r.locked || index in r.marks || index <= r.maxMarkedIndex) return false
        if (index == MixxRow.LOCK_INDEX) return r.marks.size >= 5
        return true
    }

    fun mark(rowIndex: Int, index: Int) {
        if (!canMark(rowIndex, index)) return
        var r = state.rows[rowIndex]
        r = r.copy(marks = r.marks + index)
        var didLock = false
        if (index == MixxRow.LOCK_INDEX) {
            r = r.copy(locked = true)
            didLock = true
        }
        setRow(rowIndex, r)
        recordAction(MixxAction.Mark(rowIndex, index, didLock))
    }

    // --- Penalties ---

    fun canAddPenalty(): Boolean = !isGameOver && state.penalties < MixxState.MAX_PENALTIES

    fun addPenalty() {
        if (!canAddPenalty()) return
        state = state.copy(penalties = state.penalties + 1)
        recordAction(MixxAction.Penalty)
    }

    // --- Concede a row / finish manually ---

    /**
     * You may close (concede) a row whose colour another player locked: the
     * row closes for you, but you score no lock bonus - you never crossed
     * its final cell. Allowed on any still-open row while the game is live.
     */
    fun canConcedeRow(rowIndex: Int): Boolean = !isGameOver && !state.rows[rowIndex].locked

    fun concedeRow(rowIndex: Int) {
        if (!canConcedeRow(rowIndex)) return
        val r = state.rows[rowIndex]
        setRow(rowIndex, r.copy(locked = true))
        recordAction(MixxAction.Concede(rowIndex))
    }

    fun isLastConcede(rowIndex: Int): Boolean {
        val last = state.history.lastOrNull() as? MixxAction.Concede ?: return false
        return last.row == rowIndex
    }

    /** End the game by hand - e.g. another player crossed the final lock. */
    fun canFinishManually(): Boolean = !isGameOver

    fun finishGame() {
        if (!canFinishManually()) return
        state = state.copy(manuallyFinished = true)
        recordAction(MixxAction.Finish)
    }

    // --- Scoreboard ---

    /** Crosses counted toward a row's score: its own marks plus the lock. */
    fun crosses(rowIndex: Int): Int = state.rows[rowIndex].scoringCrosses

    fun points(rowIndex: Int): Int = scoring.points(forCrosses = crosses(rowIndex))

    val penaltyPoints: Int get() = state.penalties * 5

    override val totalScore: Int
        get() = state.rows.indices.sumOf { points(it) } - penaltyPoints

    /**
     * Ends when two rows are locked, the 4th penalty is taken, or the
     * player ends it by hand.
     */
    override val isGameOver: Boolean
        get() = state.manuallyFinished || lockedRowCount >= 2 || state.penalties >= MixxState.MAX_PENALTIES

    override val canUndo: Boolean get() = state.history.isNotEmpty()

    // --- Tap-to-undo (strictly LIFO: only the most-recent action qualifies) ---

    /**
     * `true` if the most-recent action is crossing [index] in row
     * [rowIndex], so tapping that cell un-checks it (a second way to undo).
     */
    fun isLastMark(rowIndex: Int, index: Int): Boolean {
        val last = state.history.lastOrNull() as? MixxAction.Mark ?: return false
        return last.row == rowIndex && last.index == index
    }

    /** `true` if the most-recent action is the last penalty, so tapping it undoes it. */
    fun isLastPenalty(): Boolean = state.history.lastOrNull() is MixxAction.Penalty

    /**
     * Reverse the most recent action. Strictly LIFO. The undone action moves
     * onto the (in-memory only) redo stack.
     */
    override fun undo() {
        val last = state.history.lastOrNull() ?: return
        state = state.copy(history = state.history.dropLast(1))
        when (last) {
            is MixxAction.Mark -> {
                var r = state.rows[last.row]
                r = r.copy(marks = r.marks - last.index)
                if (last.didLock) r = r.copy(locked = false)
                setRow(last.row, r)
            }
            is MixxAction.Penalty -> {
                state = state.copy(penalties = maxOf(0, state.penalties - 1))
            }
            is MixxAction.Concede -> {
                val r = state.rows[last.row]
                setRow(last.row, r.copy(locked = false))
            }
            is MixxAction.Finish -> {
                state = state.copy(manuallyFinished = false)
            }
        }
        redoStack = redoStack + last
    }

    override val canRedo: Boolean get() = redoStack.isNotEmpty()

    /**
     * Re-apply the most recently undone action through the SAME mutator a
     * fresh move takes, so scores/locks/derived state stay exact - never
     * re-implement the effect here.
     */
    override fun redo() {
        val next = redoStack.lastOrNull() ?: return
        redoStack = redoStack.dropLast(1)
        isRedoing = true
        try {
            when (next) {
                is MixxAction.Mark -> mark(next.row, next.index)
                is MixxAction.Penalty -> addPenalty()
                is MixxAction.Concede -> concedeRow(next.row)
                is MixxAction.Finish -> finishGame()
            }
        } finally {
            isRedoing = false
        }
    }

    override fun reset() {
        state = MixxState()
        redoStack = emptyList()
    }

    /** Replace the engine's state wholesale, e.g. after loading a persisted snapshot. */
    fun restore(state: MixxState) {
        // Copy the collections defensively: `marks` is typed Set<Int>, so a
        // caller holding a MutableSet reference could otherwise mutate
        // engine state from outside - impossible in Swift, where rows are
        // value-semantic structs copied on assignment.
        this.state = state.copy(
            rows = state.rows.map { it.copy(marks = it.marks.toSet()) },
            history = state.history.toList(),
        )
        redoStack = emptyList()
    }

    // --- Mutation helpers ---

    /**
     * Appends a new action to the history. Any FORWARD move - i.e. every
     * call site except [redo] re-applying an undone one - invalidates the
     * redo stack (standard editor semantics: making a new move after
     * undoing forecloses the redone future).
     */
    private fun recordAction(action: MixxAction) {
        state = state.copy(history = state.history + action)
        if (!isRedoing) redoStack = emptyList()
    }

    private fun setRow(rowIndex: Int, r: MixxRow) {
        state = state.copy(rows = state.rows.toMutableList().also { it[rowIndex] = r })
    }
}
